QuteErrorPageSetup does not support templates that are not backed by a file

### Describe the bug

```
aused by: io.quarkus.qute.TemplateException: Parser error in template [roq-gen/posts/2024-10-13-the-first-roq.html] line 21: invalid virtual method in {page.date.format('YYYY MM dd'}
	at io.quarkus.qute.TemplateException$Builder.build(TemplateException.java:169)
	at io.quarkus.qute.Expressions.parseVirtualMethodParams(Expressions.java:42)
	at io.quarkus.qute.Parser.createPart(Parser.java:1049)
	at io.quarkus.qute.Parser.parseExpression(Parser.java:1035)
	at io.quarkus.qute.Parser.createExpression(Parser.java:1115)
	at io.quarkus.qute.Parser.flushTag(Parser.java:444)
	at io.quarkus.qute.Parser.tag(Parser.java:350)
	at io.quarkus.qute.Parser.processCharacter(Parser.java:251)
	at io.quarkus.qute.Parser.parse(Parser.java:158)
	at io.quarkus.qute.EngineImpl.load(EngineImpl.java:217)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1713)
	at io.quarkus.qute.EngineImpl.getTemplate(EngineImpl.java:127)
	at io.quarkus.qute.deployment.QuteProcessor.analyzeTemplates(QuteProcessor.java:755)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
	at org.jboss.threads.Enhancexecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
	at java.base/java.lang.Thread.run(Thread.java:1575)
	at org.jboss.threads.JBossThread.run(JBossThread.java:499)

2024-11-11 12:23:15,821 INFO  [io.qua.dep.dev.RuntimeUpdatesProcessor] (vert.x-worker-thread-10) Live reload total time: 0.114s
2024-11-11 12:23:15,822 WARN  [io.qua.qut.run.dev.QuteErrorPageSetup] (vert.x-eventloop-thread-1) Unable to read the template source: roq-gen/posts/2024-10-13-the-first-roq.html: java.lang.IllegalStateException: Template source not available
	at io.quarkus.qute.runtime.devmode.QuteErrorPageSetup.getBufferedReader(QuteErrorPageSetup.java:190)
	at io.quarkus.qute.runtime.devmode.QuteErrorPageSetup.getProblemInfo(QuteErrorPageSetup.java:133)
	at io.quarkus.qute.runtime.devmode.QuteErrorPageSetup.generatePage(QuteErrorPageSetup.java:115)
	at io.quarkus.vertx.http.runtime.devmode.ReplacementDebugPage.generateHtml(ReplacementDebugPage.java:32)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup.handleDeploymentProblem(VertxHttpHotReplacementSetup.java:213)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$5.handle(VertxHttpHotReplacementSetup.java:196)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$5.handle(VertxHttpHotReplacementSetup.java:192)
	at io.vertx.core.impl.future.FutureImpl$4.onFailure(FutureImpl.java:188)
	at io.vertx.core.impl.future.FutureBase.lambda$emitFailure$1(FutureBase.java:75)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1575)


2024-11-11 12:23:15,822 ERROR [io.qua.ver.cor.run.VertxCoreRecorder] (vert.x-eventloop-thread-1) Uncaught exception received by Vert.x: io.quarkus.qute.TemplateException
	at io.quarkus.qute.CompletedStage.get(CompletedStage.java:65)
	at io.quarkus.qute.IfSectionHelper$DoubletonCondition.evaluate(IfSectionHelper.java:294)
	at io.quarkus.qute.IfSectionHelper$SingletonContext.resolve(IfSectionHelper.java:64)
	at io.quarkus.qute.IfSectionHelper.resolve(IfSectionHelper.java:45)
	at io.quarkus.qute.SectionNode.resolve(SectionNode.java:53)
	at io.quarkus.qute.SectionNode.resolve(SectionNode.java:58)
	at io.quarkus.qute.Results.resolveWith(Results.java:110)
	at io.quarkus.qute.Results.resolveAndProcess(Results.java:57)
	at io.quarkus.qute.SectionNode$SectionResolutionContextImpl.execute(SectionNode.java:238)
	at io.quarkus.qute.SectionHelper$SectionResolutionContext.execute(SectionHelper.java:66)
	at io.quarkus.qute.Parser$1.resolve(Parser.java:1331)
	at io.quarkus.qute.SectionNode.resolve(SectionNode.java:53)
	at io.quarkus.qute.SectionNode.resolve(SectionNode.java:58)
	at io.quarkus.qute.TemplateImpl$TemplateInstanceImpl.renderData(TemplateImpl.java:272)
	at io.quarkus.qute.TemplateImpl$TemplateInstanceImpl.renderAsyncNoTimeout(TemplateImpl.java:239)
	at io.quarkus.qute.TemplateImpl$TemplateInstanceImpl.render(TemplateImpl.java:164)
	at io.quarkus.qute.runtime.devmode.QuteErrorPageSetup.getProblemInfo(QuteErrorPageSetup.java:171)
	at io.quarkus.qute.runtime.devmode.QuteErrorPageSetup.generatePage(QuteErrorPageSetup.java:115)
	at io.quarkus.vertx.http.runtime.devmode.ReplacementDebugPage.generateHtml(ReplacementDebugPage.java:32)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup.handleDeploymentProblem(VertxHttpHotReplacementSetup.java:213)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$5.handle(VertxHttpHotReplacementSetup.java:196)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$5.handle(VertxHttpHotReplacementSetup.java:192)
	at io.vertx.core.impl.future.FutureImpl$4.onFailure(FutureImpl.java:188)
	at io.vertx.core.impl.future.FutureBase.lambda$emitFailure$1(FutureBase.java:75)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1575)
Caused by: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
	at java.base/java.util.Objects.checkIndex(Objects.java:365)
	at java.base/java.util.ArrayList.get(ArrayList.java:428)
	at io.quarkus.qute.ValueResolvers.lambda$listResolveAsync$1(ValueResolvers.java:631)
	at io.quarkus.qute.CompletedStage.thenApply(CompletedStage.java:76)
	at io.quarkus.qute.ValueResolvers.listResolveAsync(ValueResolvers.java:628)
	at io.quarkus.qute.ValueResolvers$ListResolver.resolve(ValueResolvers.java:64)
	at io.quarkus.qute.EvaluatorImpl.resolve(EvaluatorImpl.java:211)
	at io.quarkus.qute.EvaluatorImpl.resolveReference(EvaluatorImpl.java:131)
	at io.quarkus.qute.EvaluatorImpl.lambda$resolveReference$2(EvaluatorImpl.java:135)
	at io.quarkus.qute.CompletedStage.thenCompose(CompletedStage.java:249)
	at io.quarkus.qute.EvaluatorImpl.resolveReference(EvaluatorImpl.java:135)
	at io.quarkus.qute.EvaluatorImpl.evaluate(EvaluatorImpl.java:85)
	at io.quarkus.qute.ResolutionContextImpl.evaluate(ResolutionContextImpl.java:29)
	at io.quarkus.qute.IfSectionHelper$OperandCondition.evaluate(IfSectionHelper.java:253)
	at io.quarkus.qute.IfSectionHelper$DoubletonCondition.evaluate(IfSectionHelper.java:292)
	... 30 more


2024-11-11 12:23:56,724 WARN  [io.net.boo.ServerBootstrap] (vert.x-acceptor-thread-0) Failed to register an accepted channel: [id: 0x4b7ea862, L:/127.0.0.1:8080 ! R:/127.0.0.1:54053]: java.lang.IllegalStateException
	at io.vertx.core.net.impl.VertxEventLoopGroup.next(VertxEventLoopGroup.java:37)
	at io.vertx.core.net.impl.VertxEventLoopGroup.register(VertxEventLoopGroup.java:53)
	at io.netty.bootstrap.ServerBootstrap$ServerBootstrapAcceptor.channelRead(ServerBootstrap.java:241)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1407)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:918)
	at io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:97)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:553)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1575)
```


### How to Reproduce?

Create Roq app:
```shell
quarkus create app roq-with-blog -x=io.quarkiverse.roq:quarkus-roq
```


open dev mode, edit content/posts/2024-10-13-the-first-roq.md by removing the closing `)` from the `page.date.format('YYYY MM dd'`


_No response_

--- a/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteDevModeProcessor.java
+++ b/extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteDevModeProcessor.java
@@ -1,0 +1,31 @@
+package io.quarkus.qute.deployment;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import io.quarkus.arc.deployment.ValidationPhaseBuildItem.ValidationErrorBuildItem;
+import io.quarkus.deployment.IsDevelopment;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.annotations.BuildSteps;
+import io.quarkus.dev.console.DevConsoleManager;
+import io.quarkus.qute.runtime.devmode.QuteErrorPageSetup;
+
+@BuildSteps(onlyIf = IsDevelopment.class)
+public class QuteDevModeProcessor {
+
+    @BuildStep
+    void collectGeneratedContents(List<TemplatePathBuildItem> templatePaths,
+            BuildProducer<ValidationErrorBuildItem> errors) {
+        Map<String, String> contents = new HashMap<>();
+        for (TemplatePathBuildItem template : templatePaths) {
+            if (!template.isFileBased()) {
+                contents.put(template.getPath(), template.getContent());
+            }
+        }
+        // Set the global that could be used at runtime when a qute error page is rendered
+        DevConsoleManager.setGlobal(QuteErrorPageSetup.GENERATED_CONTENTS, contents);
+    }
+
+}

--- a/extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/devmode/QuteErrorPageSetup.java
+++ b/extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/devmode/QuteErrorPageSetup.java
@@ -2,6 +2,7 @@ package io.quarkus.qute.runtime.devmode;
 
 import java.io.BufferedReader;
 import java.io.IOException;
+import java.io.StringReader;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.nio.file.Files;
@@ -12,6 +13,7 @@ import java.util.Collections;
 import java.util.Comparator;
 import java.util.List;
 import java.util.ListIterator;
+import java.util.Map;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
 import java.util.stream.Collectors;
@@ -19,6 +21,7 @@ import java.util.stream.Collectors;
 import org.jboss.logging.Logger;
 
 import io.quarkus.dev.ErrorPageGenerators;
+import io.quarkus.dev.console.DevConsoleManager;
 import io.quarkus.dev.spi.HotReplacementContext;
 import io.quarkus.dev.spi.HotReplacementSetup;
 import io.quarkus.qute.Engine;
@@ -32,6 +35,8 @@ import io.quarkus.runtime.TemplateHtmlBuilder;
 public class QuteErrorPageSetup implements HotReplacementSetup {
 
     private static final Logger LOG = Logger.getLogger(QuteErrorPageSetup.class);
+
+    public static final String GENERATED_CONTENTS = "io.quarkus.qute.generatedContents";
 
     private static final String TEMPLATE_EXCEPTION = "io.quarkus.qute.TemplateException";
     private static final String ORIGIN = "io.quarkus.qute.TemplateNode$Origin";
@@ -139,6 +144,10 @@ public class QuteErrorPageSetup implements HotReplacementSetup {
             LOG.warn("Unable to read the template source: " + templateId, e);
         }
 
+        if (sourceLines.isEmpty()) {
+            return Arrays.stream(messageLines).collect(Collectors.joining("<br>"));
+        }
+
         List<Integer> realLines = new ArrayList<>();
         boolean endLinesSkipped = false;
         if (sourceLines.size() > 15) {
@@ -185,6 +194,14 @@ public class QuteErrorPageSetup implements HotReplacementSetup {
                 if (Files.exists(template)) {
                     return Files.newBufferedReader(template);
                 }
+            }
+        }
+        // Source file not available - try to search the generated contents
+        Map<String, String> generatedContents = DevConsoleManager.getGlobal(GENERATED_CONTENTS);
+        if (generatedContents != null) {
+            String template = generatedContents.get(templateId);
+            if (template != null) {
+                return new BufferedReader(new StringReader(template));
             }
         }
         throw new IllegalStateException("Template source not available");

